src/Haskell/Ide/Engine/Server.hs

```diff
@@ -151,12 +152,64 @@ run scheduler _origDir plugins captureFp = flip E.catches handlers $ do
+        let dummyCradleFile = fromMaybe currentDir lspRootDir </> "File.hs"
+        debugm $ "Dummy Cradle file result: " ++ dummyCradleFile
+        cradleRes <- liftIO $ E.try (findLocalCradle dummyCradleFile)
+        let sf = Core.sendFunc lf
+
+        case cradleRes of
+          Right cradle -> do
+            projGhcVersion <- liftIO $ getProjectGhcVersion cradle
+            when (projGhcVersion /= hieGhcVersion) $ do
+              let msg = T.pack $ "Mismatching GHC versions: " ++ cradleDisplay cradle ++
+                        " is " ++ projGhcVersion ++ ", HIE is " ++ hieGhcVersion
+                        ++ "\nYou may want to use hie-wrapper. Check the README for more information"
+              sf $ NotShowMessage $ fmServerShowMessageNotification J.MtWarning msg
+              sf $ NotLogMessage $ fmServerLogMessageNotification J.MtWarning msg
+
+              -- Check cabal is installed
+              when (isCabalCradle cradle) $ do
+                hasCabal <- liftIO checkCabalInstall
+                unless hasCabal $ do
+                  let cabalMsg = T.pack "cabal-install is not installed. Check the README for more information"
+                  sf $ NotShowMessage $ fmServerShowMessageNotification J.MtWarning cabalMsg
+                  sf $ NotLogMessage $ fmServerLogMessageNotification J.MtWarning cabalMsg
+
+          Left (e :: Yaml.ParseException) -> do
+            logm $ "Failed to parse `hie.yaml`: " ++ show e
+            sf $ NotShowMessage $ fmServerShowMessageNotification J.MtError ("Couldn't parse hie.yaml: \n" <> T.pack (show e))
+
+        let mcradle = case cradleRes of
+              Left _ -> Nothing
+              Right c -> Just c
+
         -- haskell lsp sets the current directory to the project root in the InitializeRequest
         -- We launch the dispatcher after that so that the default cradle is
         -- recognized properly by ghc-mod
-        flip labelThread "scheduler" =<< (forkIO $ Scheduler.runScheduler scheduler errorHandler callbackHandler (Just lf))
-        flip labelThread "reactor" =<< (forkIO reactorFunc)
-        flip labelThread "diagnostics" =<< (forkIO $ diagnosticsQueue tr)
+        flip labelThread "scheduler" =<<
+            (forkIO (
+              Scheduler.runScheduler scheduler errorHandler callbackHandler (Just lf) mcradle
+              `E.catch` \(e :: E.SomeException) ->
+              (errorm $ "Scheduler thread exited unexpectedly: " ++ show e)
+            ))
+        flip labelThread "reactor" =<<
+            (forkIO (
+              reactorFunc
+              `E.catch` \(e :: E.SomeException) ->
+              (errorm $ "Reactor thread exited unexpectedly: " ++ show e)
+            ))
+        flip labelThread "diagnostics" =<<
+            (forkIO (
+              diagnosticsQueue tr
+              `E.catch` \(e :: E.SomeException) ->
+              (errorm $ "Diagnostic thread exited unexpectedly: " ++ show e)
+            ))
+
```

Comment:
At some future date we should go back to launching these threads with `race` rather than `forkIO`, so if one does die, we know immediately, and can see something in the log.  With this setup it can still limp along.